objects/src/block/mod.rs

```diff
@@ -146,26 +152,41 @@ impl Block {
     // --------------------------------------------------------------------------------------------
 
     fn validate(&self) -> Result<(), BlockError> {
+        let account_count = self.updated_accounts.len();
+        if account_count > MAX_ACCOUNTS_PER_BLOCK {
+            return Err(BlockError::TooManyAccountsUpdated(account_count));
+        }
+
         let batch_count = self.output_note_batches.len();
         if batch_count > MAX_BATCHES_PER_BLOCK {
             return Err(BlockError::TooManyTransactionBatches(batch_count));
         }
 
-        for batch in self.output_note_batches.iter() {
-            if batch.len() > MAX_NOTES_PER_BATCH {
-                return Err(BlockError::TooManyNotesInBatch(batch.len()));
-            }
+        // We can't check input notes here because they're not stored in the block,
+        // so we check that nullifier count is not bigger than maximum input notes per block.
+        let nullifier_count = self.nullifiers.len();
+        if nullifier_count > MAX_INPUT_NOTES_PER_BLOCK {
+            return Err(BlockError::TooManyNullifiersInBlock(nullifier_count));
         }
 
-        let mut notes = BTreeSet::new();
+        let mut output_notes = BTreeSet::new();
+        let mut output_note_count = 0;
         for batch in self.output_note_batches.iter() {
+            if batch.len() > MAX_OUTPUT_NOTES_PER_BATCH {
+                return Err(BlockError::TooManyNotesInBatch(batch.len()));
+            }
```

Comment:
not for this PR, but I think we should construct blocks from something like `TransactionBatch` structs. Then, we'd check that transaction batches have the right number of nullifiers/output notes and we won't need to do such checks here. Let's leave a `TODO` here about this.